File: C++/20.ValidParentheses.cpp
/*
    Title:          20.ValidParentheses.cpp
    Update date:    2019/11/06
    Author:         Zhuofan Zhang
*/
class Solution {
public:
    bool isValid(string s) 
    {
        /* Method1:(100.00%,96.12%) */
        stack<char> parse_stack;
        for(auto &p: s)
        {
            switch(p)
            {
                case '(':parse_stack.push('(');break;
                case '[':parse_stack.push('[');break;
                case '{':parse_stack.push('{');break;
                case ')':
                    if( parse_stack.empty() || parse_stack.top()!='(' )
                        return false;
                    else
                        parse_stack.pop();
                    break;
                case ']':
                    if( parse_stack.empty() || parse_stack.top()!='[' )
                        return false;
                    else
                        parse_stack.pop();
                    break;
                case '}':
                    if( parse_stack.empty() || parse_stack.top()!='{' )
                        return false;
                    else
                        parse_stack.pop();
                    break;
                default:;
            }
              
        }
        
        return parse_stack.empty();
    }
};

File: Golang/Stack/20.ValidParentheses.go
/*
    Title:          20.ValidParentheses.go
    Update date:    2020/09/21
    Author:         Zhuofan Zhang
*/
package solution

// --------- stack impl ---------

type stack struct {
    Container []rune
}

func (s *stack) top() rune {
    if l := len(s.Container); l != 0 {
        return s.Container[l-1]
    } else {
        panic("top from empty stack.")
    }
}

func (s *stack) push(c rune) {
    s.Container = append(s.Container, c)
}

func (s *stack) pop() {
    if l := len(s.Container); l != 0 {
        s.Container = s.Container[:l-1]
    } else {
        panic("pop from empty stack.")
    }
}

func (s *stack) isEmpty() bool {
    return len(s.Container) == 0
}

// -------------------------------

func isValid(s string) bool {
    var stk stack
    for _, c := range s {
        switch ch := c; ch {
            case '(': stk.push(c)
            case '[': stk.push(c)
            case '{': stk.push(c)
            case ')':
                if stk.isEmpty() || stk.top() != '(' {
                    return false
                } else {
                    stk.pop()
                }
            case ']':
                if stk.isEmpty() || stk.top() != '[' {
                    return false
                } else {
                    stk.pop()
                }
            case '}':
                if stk.isEmpty() || stk.top() != '{' {
                    return false
                } else {
                    stk.pop()
                }
        }
    }

    return stk.isEmpty()
}